MergeSharpWebAPI/Services/GraphService.cs

```diff
@@ -0,0 +1,153 @@
+using System.Text.Json.Serialization;
+using MergeSharpWebAPI.Models;
+
+namespace MergeSharpWebAPI.Services;
+
+// The service should act as the translator between frontend key and backend Guid
+public class GraphService
+{
+    private readonly Graph _graph;
+    private readonly Dictionary<Guid, int> _vertexGuidToKeyMap;
+    private readonly Dictionary<int, Graph.Vertex> _keyToVertexMap;
+
+    public GraphService()
+    {
+        _graph = new();
+        _vertexGuidToKeyMap = new();
+        _keyToVertexMap = new();
+    }
+
+    // Json object for the frontend
+    public readonly struct VertexInfo
+    {
+        [JsonInclude]
+        public readonly int key;
+        [JsonInclude]
+        public readonly double x;
+        [JsonInclude]
+        public readonly double y;
+        [JsonInclude]
+        public readonly string type;
+
+        public VertexInfo(int key, double x, double y, Graph.Vertex.Type type)
+        {
+            this.key = key;
+            this.x = x;
+            this.y = y;
+            this.type = type.ToString().ToLower();
+        }
+    }
+
+    public Dictionary<(int, int), int> EdgeCounts => _graph.EdgeCounts().ToDictionary(
+                                                                kv => (_vertexGuidToKeyMap[kv.Key.src], _vertexGuidToKeyMap[kv.Key.dst]),
+                                                                kv => kv.Value
+                                                            );
+
+    public int EdgeCount(int srcKey, int dstKey)
+    {
+        if (_keyToVertexMap.TryGetValue(srcKey, out Graph.Vertex src)
+                && _keyToVertexMap.TryGetValue(dstKey, out Graph.Vertex dst))
+        {
+            return _graph.EdgeCount(new Graph.Edge(src.guid, dst.guid));
+        }
+
+        return 0;
+    }
+
+    public IEnumerable<VertexInfo> Vertices => _keyToVertexMap.Select(
+        (kv) =>
+        {
+            int key = kv.Key;
+            Graph.Vertex vertex = kv.Value;
+            return new VertexInfo(key, vertex.x, vertex.y, vertex.type);
+        }
+    );
+    public VertexInfo Vertex(int key)
+    {
+        if (_keyToVertexMap.TryGetValue(key, out Graph.Vertex v))
+        {
+            return new VertexInfo(key, v.x, v.y, v.type);
+        }
+
+        // TODO: throw an exception
+        return new VertexInfo();
+    }
+
+    public bool AddVertex(int key, double x, double y, string stype)

```

Comment:
just curious why you named the variable stype, is this a naming convention 0.0

Reply:
Nah just a for string cause there's the string type and then the Type type

Reply:
Gotchu makes sense